agent/metrics/metrics.go

```diff
@@ -0,0 +1,214 @@
+        monitoredrespb "google.golang.org/genproto/googleapis/api/monitoredres"
+)
+
+const (
+	samplePeriod	= 60 * time.Second
+        endpoint	= "monitoring.sandbox.googleapis.com:443"
+)
+
+var startTime time.Time
+var codeCount map[string]int64
+
+// metricClient is a client for interacting with Cloud Monitoring API.
+type metricClient interface {
+	CreateTimeSeries(ctx context.Context, req *monitoringpb.CreateTimeSeriesRequest, opts ...gax.CallOption) error
+}
+
+// fakeMetricClient is a stub of MetricClient for the purpose of testing
+type fakeMetricClient struct {
+	Requests []*monitoringpb.CreateTimeSeriesRequest
+}
+
+func (f *fakeMetricClient) CreateTimeSeries(ctx context.Context, req *monitoringpb.CreateTimeSeriesRequest, opts ...gax.CallOption) error {
+	f.Requests = append(f.Requests, req)
+	return nil
+}
+
+type MetricHandler struct {
+	projectID string
+	instanceID string
+	instanceZone string
+	metricDomain string
+	ctx context.Context
+	client metricClient
+}
+
+// NewFakeMetricHandler instantiates a fake metric client for the purpose of testing
+func NewFakeMetricHandler(ctx context.Context, projectID, instanceID, instanceZone, metricDomain string) (*MetricHandler, error) {
+	client := fakeMetricClient{}
+	return newMetricHandlerHelper(ctx, projectID, instanceID, instanceZone, metricDomain, &client)
+}
+
+// NewMetricHandler instantiates a metric client for the purpose of writing metrics to cloud monarch
+func NewMetricHandler(ctx context.Context, projectID, instanceID, instanceZone, metricDomain string) (*MetricHandler, error) {
+	log.Printf("NewMetricHandler|instantiating metric handler")
+	client, err := monitoring.NewMetricClient(
+                ctx,
+                apioption.WithEndpoint(endpoint),
+        )
+	if err != nil {
+		log.Fatalf("Failed to create client: %v", err)
+		return nil, err
+	}
+	return newMetricHandlerHelper(ctx, projectID, instanceID, instanceZone, metricDomain, client)
+}
+
+func newMetricHandlerHelper(ctx context.Context, projectID, instanceID, instanceZone, metricDomain string, client metricClient) (*MetricHandler, error) {
+	if projectID == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing projectID")
+		return nil, err
+	}
+	if instanceID == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing instanceID")
+		return nil, err
+	}
+	if instanceZone == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing instanceZone")
+		return nil, err
+	}
+	if metricDomain == "" {
+		err := fmt.Errorf("Failed to create metric handler: missing metricDomain")
+		return nil, err
+	}
+
+	startTime = time.Now()
+	codeCount = make(map[string]int64)
+
+	return &MetricHandler{
+		projectID:	projectID,
+		instanceID:	instanceID,
+		instanceZone:	instanceZone,
+		metricDomain:	metricDomain,
+		ctx:		ctx,
+		client:		client,
+	}, nil
+}
+
+func (h *MetricHandler) WriteMetric(metricType string, statusCode int) error {
```

Comment:
I don't understand the purpose of this method.

It seems like the client code would be both shorter and clearer if it directly called `writeResponseCodeMetric`.

Let's just capitalize that method name so it is exported by the package, and then have clients directly call it.

Reply:
Originally, I was thinking about introducing another metric that tracks just the failure responses, so hence this function, but you're right - it's not needed at this time. Removed.